### PR TITLE
[misc] Slient & non-blocking version check

### PR DESCRIPTION
**Pull request by Leonz5288**
Now version check runs silently per 7 days. Nothing will be printed in terminal.
If user's version is outdated, a latest_version file will be written to ~/.taichi and the latest version will be recorded.
After that, no more version check in 7 days. When user use Taichi again, if the current version is lower than the version recorded in the file, the latest version will be printed with the Taichi header. For example:
`[Taichi] version 0.8.7, latest version 0.9.0, llvm 10.0.0, commit abcdef, linux, python 3.8.10`

### CODE REVIEW

**Comment by Leonz5288** [comment]
Now this service runs in a new thread and main process will not wait for the new thread to finish (prevent blocking user).
I also tested a few cases related to network issues, for example, timeout, domain mismatch, network down, etc. Nothing will be printed on terminal and Taichi main process will run as normal.
If there are other cases that comes to your mind, please tell me so I can test more.

File: python/taichi/lang/__init__.py [hunk 1]
```diff
@@ -453,18 +453,15 @@ def check_version():
...
-            elif response['status'] == 0:
-                # Status 0 means that user already have the latest Taichi. The message here prompts this infomation to users.
-                print(response['message'])
-    except Exception as error:
-        print('Checking lastest version failed:', error)
+                version_path = os.path.join(_ti_core.get_repo_dir(),
+                                            'latest_version')
+                with open(version_path, 'w') as f:
+                    f.write(response['latest_version'])
+    except Exception:
```
Comment (ailzhang) [inline python/taichi/lang/__init__.py:463]: Wild but might be a good fit here as we discussed @rexwangcc @strongoier :P  : 
```
except:
    pass
```

File: python/taichi/lang/__init__.py [hunk 2]
```diff
@@ -481,15 +478,14 @@ def try_check_version():
...
-                    f.truncate()
         else:
             check_version()
             with open(timestamp_path, 'w') as f:
                 f.write((cur_date +
                          datetime.timedelta(days=7)).strftime('%Y-%m-%d'))
     # Wildcard exception to catch potential file writing errors.
-    except Exception as error:
-        print('Checking lastest version failed:', error)
+    except Exception:
```
Comment (ailzhang) [inline python/taichi/lang/__init__.py:487]: ditto

File: python/taichi/lang/__init__.py [hunk 3]
```diff
@@ -521,7 +517,10 @@ def init(arch=None,
     # Check version for users every 7 days if not disabled by users.
     skip = os.environ.get("TI_SKIP_VERSION_CHECK")
     if skip != 'ON':
-        try_check_version()
+        # We don't join this thread because we do not wish to block the user.
+        check_version_thread = threading.Thread(target=try_check_version,
```
Comment (ailzhang) [inline python/taichi/lang/__init__.py:521]: Just to double check: have you tried locally that just raise in this thread and see if it crash the main process?

**Review by ailzhang — COMMENTED** [review]
LGTM!

File: python/taichi/lang/__init__.py (hunk 1 above)
Comment (Leonz5288) [reply python/taichi/lang/__init__.py:463]: OK, didn't know it can work like this

File: python/taichi/lang/__init__.py (hunk 3 above)
Comment (Leonz5288) [reply python/taichi/lang/__init__.py:521]: I tried crash the thread and not catch error. The sub thread will crash and a bunch of output will be printed, but the main thread can run as normal.

File: python/taichi/lang/__init__.py [hunk 4]
```diff
@@ -481,15 +478,14 @@ def try_check_version():
                 with open(timestamp_path, 'w') as f:
```
Comment (frostming) [inline python/taichi/lang/__init__.py:478]: We may not need a timestamp file. Reading the mtime should work.

Comment (Leonz5288) [reply python/taichi/lang/__init__.py:478]: What do you mean by mtime?

File: python/taichi/lang/__init__.py (hunk 3 above)
Comment (frostming) [reply python/taichi/lang/__init__.py:521]: Yeah, the exceptions raised in sub thread can't fail the main thread. But I am not sure if `sys.exit` would work

File: python/taichi/lang/__init__.py (hunk 4 above)
Comment (frostming) [reply python/taichi/lang/__init__.py:478]: last modified time returned by `os.stat(filename)`

File: python/taichi/_lib/utils.py [hunk 5]
```diff
@@ -142,10 +142,33 @@ def require_version(major, minor=None, patch=None):
 at_startup()
 
 
+def compare_version(latest, current):
+    latest_num = map(int, latest.split('.'))
+    current_num = map(int, current.split('.'))
+    for x, y in zip(latest_num, current_num):
```
Comment (frostming) [inline python/taichi/_lib/utils.py:148]: The `tuple[int]` can be compared directly with `<`/`>`/`==`

Comment (Leonz5288) [reply python/taichi/_lib/utils.py]: Thanks for the info!

File: python/taichi/lang/__init__.py (hunk 4 above)
Comment (Leonz5288) [reply python/taichi/lang/__init__.py:478]: I see. However, I don't think we can guarantee that users won't modify the file. Maybe we should stay timestamp for now?

Comment (frostming) [reply python/taichi/lang/__init__.py:478]: Reasonable, but can we store timestamp with the version in the same file, maybe a json file for convenience. In this way we don't scatter misc files around.

File: python/taichi/lang/__init__.py (hunk 1 above)
Comment (rexwangcc) [reply python/taichi/lang/__init__.py]: didn't know it either, tested with:

```python
try:
     raise BaseException(99)
except:
     print(42)
```

**Review by rexwangcc — APPROVED** [review]
LGTM, thanks! (let's resolve all comments before merging :smile: )

File: python/taichi/lang/__init__.py (hunk 4 above)
Comment (Leonz5288) [reply python/taichi/lang/__init__.py:478]: Understood, let me adjust this.

File: python/taichi/lang/__init__.py (hunk 3 above)
Comment (Leonz5288) [reply python/taichi/lang/__init__.py:521]: Can you elaborate about sys.exit please?

Comment (frostming) [reply python/taichi/lang/__init__.py:521]: I may have understood your need, if you don't want it to fail the main program, then subthread is OK.

Comment (Leonz5288) [reply python/taichi/lang/__init__.py:531]: Yes, the main program is the last thing we wish to fail.

**Pull request merged**